Applications/Tizen_native/CustomShortcut/src/data.c

```diff
@@ -43,6 +43,200 @@ int parse_route(const char *source, char **route, char **data) {
   return APP_ERROR_NONE;
 }
 
+void data_get_resource_path(const char *file, char *full_path, bool shared) {
+  char *root_path;
+  if (shared) {
+    root_path = app_get_shared_resource_path();
+  } else {
+    root_path = app_get_resource_path();
+  }
+
+  if (root_path == NULL) {
+    LOG_E("failed to get resource path");
+    return;
+  }
+
+  if (full_path == NULL) {
+    LOG_E("full_path is null");
+  }
+
+  snprintf(full_path, PATH_MAX, "%s%s", root_path, file);
+  LOG_D("resource path: %s", full_path);
+  free(root_path);
+}
+
+void data_get_data_path(const char *file, char *full_path) {
+  char *root_path;
+
+  root_path = app_get_data_path();
+
+  if (root_path == NULL) {
+    LOG_E("failed to get data path");
+    return;
+  }
+
+  if (full_path == NULL) {
+    LOG_E("full_path is null");
+    return;
+  }
+
+  snprintf(full_path, PATH_MAX, "%s%s", root_path, file);
+  LOG_D("data path: %s", full_path);
+  free(root_path);
+}
+
+static void _on_data_receive(ml_tensors_data_h data,
+                             const ml_tensors_info_h info, void *user_data) {
+
+  appdata_s *ad = (appdata_s *)user_data;
+
+  void *raw_data;
+  size_t data_size;
+  FILE *file;
+
+  int status = ml_tensors_data_get_tensor_data(data, 0, &raw_data, &data_size);
+  LOG_D("size of data: %ud", data_size);
+
+  if (status != ML_ERROR_NONE) {
+    LOG_E("get tensor data failed %d", status);
+    return;
+  }
+
+  status = ml_tensors_info_get_tensor_size(info, -1, &data_size);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("get tensor data failed %d", status);
+    return;
+  }
+
+  file = fopen(ad->pipe_dst, ad->tries == 1 ? "wb" : "wb+");
+
+  if (file == NULL) {
+    LOG_E("cannot open file");
+  }
+
+  if (fwrite(raw_data, data_size, 1, file) < 0) {
+    LOG_E("write error happend");
+  };
+
+  if (fclose(file) < 0) {
+    LOG_E("there was error closing");
+    return;
+  }
+
+  LOG_D("using pipeline finished, destroying pipeline");
+
+  ad->is_pipe_used = false;
+  LOG_D("destroying pipeline start");
+}
+
+static int _run_nnpipeline(appdata_s *ad, const char *src, bool clean) {
+  char pipe_description[5000];
+
+  char model_path[PATH_MAX];
+
+  int status = ML_ERROR_NONE;
+
+  data_get_resource_path("mobilenetv2.tflite", model_path, false);
+
+  if (ad->is_pipe_used == true)
+    LOG_W("pipe is currently used, wait until using pipe is done.");
+
+  time_t start = time(NULL);
+  while (ad->is_pipe_used) {
```

Comment:
what is the purpose this? This looks like a CPU hogger.

Reply:
Is this waiting for a previously running pipeline to execute?

Reply:
Because the pipeline is writing to a file, we have to run this synchronously, so this is waiting for other pipe to be finished (and close the file)

I couldn't find sth like p.close() for ml api

Do you have any better idea?

Reply:
Add mutex lock over it. Lock would ensure single access with sleep than utilizing the CPU.

Reply:
Mutex for now is fine, but semaphores would have been much more cleaner.
I didn't realize that there is just 1 thread always.